src/value/borrowed.rs

```diff
@@ -249,84 +250,126 @@ where
 {
     /// Returns a borrowed struct from the given iterator of field names/values.
     fn from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
-        let mut fields: LinkedHashMap<SymbolRef<'val>, SmallVec<[ElementRef<'val>; 1]>> =
-            LinkedHashMap::new();
-        let mut number_of_fields: usize = 0;
+        let mut by_index = Vec::new();
+        let mut by_name = HashMap::new();
+        for (field_name, field_value) in iter {
+            let field_name = field_name.into();
+            let field_value = field_value.into();
+
+            by_name
+                .entry(field_name.clone())
+                .or_insert_with(IndexVec::new)
+                .push(by_index.len());
+            by_index.push((field_name, field_value));
+        }
 
-        for (k, v) in iter {
-            let key = k.into();
-            let val = v.into();
+        let fields = Rc::new(FieldRefs { by_index, by_name });
+        Self { fields }
+    }
+}
 
-            fields.entry(key).or_insert_with(SmallVec::new).push(val);
+// This collection is broken out into its own type to allow instances of it to be shared with Rc.
+#[derive(Debug)]
+struct FieldRefs<'a> {
+    // Key/value pairs in the order they were inserted
+    by_index: Vec<(SymbolRef<'a>, ElementRef<'a>)>,
+    // Maps symbols to a list of indexes where values may be found in `by_index` above
+    by_name: HashMap<SymbolRef<'a>, IndexVec>,
+}
 
-            number_of_fields += 1;
+impl<'data> FieldRefs<'data> {
+    /// Gets all of the indexes that contain a value associated with the given field name.
+    fn get_indexes<A: AsSymbolRef>(&self, field_name: A) -> Option<&IndexVec> {
+        match field_name.as_symbol_ref().text() {
+            // If the provided field name symbol has undefined text...
+            None => {
+                // ...then build a cheap, stack-allocated `Symbol` that represents unknown text
+                let symbol = SymbolRef::with_unknown_text();
+                // ...and use the unknown text symbol to look up matching field values
```

Comment:
I'm trying to reason about what the behavior of this would be. What fields would match the `unknown text symbol`?

Reply:
Fields whose symbol ID intentionally maps to unknown text. This is the case for `$0` as well as any symbol whose entry in the symbol table was non-text (`null`, `5`, `false`, etc).